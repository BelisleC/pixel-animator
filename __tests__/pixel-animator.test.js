const PixelAnimator = require("../pixel-animator");

it("should iterate over frames", () => {
  const frameIterator = PixelAnimator({
    columns: 2,
    rows: 2,
    frames: 2,
    colorize: ({ column, row, frame, columns, rows, frames }) => ({
      red: column / columns,
      green: row / rows,
      blue: frame / frames
    })
  });
  const frame1 = frameIterator();
  const frame2 = frameIterator();
  const frame3 = frameIterator();
  expect([frame1, frame2]).toMatchSnapshot();
  expect(frame1).toEqual(frame3);
});

it("should generate a stateless 2×2×2 animation", () => {
  document.body.innerHTML = `<div id="root"></div>`;
  const rootElement = document.getElementById("root");
  const transport = PixelAnimator(
    {
      columns: 2,
      rows: 2,
      frames: 2,
      colorize: ({ column, row, frame, columns, rows, frames }) => ({
        red: column / columns,
        green: row / rows,
        blue: frame / frames
      })
    },
    rootElement
  );
  transport.pause();
  transport.next();
  expect(rootElement).toMatchSnapshot("first frame");
  transport.next();
  expect(rootElement).toMatchSnapshot("second frame");
});

it("should generate a stateful or evolving 2×2×2 animation", () => {
  document.body.innerHTML = `<div id="root"></div>`;
  const rootElement = document.getElementById("root");
  const transport = PixelAnimator(
    {
      columns: 2,
      rows: 2,
      frames: 2,
      frameRate: 1,
      evolve: ({ column, row, frame, cells }) => {
        if (frame === 0) return false;
        return !cells(column, row);
      },
      colorize: cell => (cell ? 1 : 0)
    },
    rootElement
  );
  transport.pause();
  transport.next();
  expect(rootElement).toMatchSnapshot("first frame");
  transport.next();
  expect(rootElement).toMatchSnapshot("second frame");
});
