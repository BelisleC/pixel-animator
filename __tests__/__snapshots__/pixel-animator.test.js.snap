// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate HTML when given a DOM node: first frame 1`] = `
<div
  class=" pixel-animator"
  id="root"
  style="width: 320px; height: 320px;"
>
  <div
    class="pa-cell pa-cell-0"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(0, 0, 0);"
  />
  <div
    class="pa-cell pa-cell-1"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(127, 0, 0);"
  />
  <div
    class="pa-cell pa-cell-2"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(0, 127, 0);"
  />
  <div
    class="pa-cell pa-cell-3"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(127, 127, 0);"
  />
</div>
`;

exports[`should generate HTML when given a DOM node: second frame 1`] = `
<div
  class=" pixel-animator"
  id="root"
  style="width: 320px; height: 320px;"
>
  <div
    class="pa-cell pa-cell-0"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(0, 0, 127);"
  />
  <div
    class="pa-cell pa-cell-1"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(127, 0, 127);"
  />
  <div
    class="pa-cell pa-cell-2"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(0, 127, 127);"
  />
  <div
    class="pa-cell pa-cell-3"
    style="width: 50%; height: 50%; display: inline-block; vertical-align: top; background-color: rgb(127, 127, 127);"
  />
</div>
`;

exports[`should iterate over frames 1`] = `
Array [
  Array [
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 127,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 127,
    },
  ],
  Array [
    Object {
      "alpha": 1,
      "blue": 127,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 127,
      "green": 0,
      "red": 127,
    },
    Object {
      "alpha": 1,
      "blue": 127,
      "green": 127,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 127,
      "green": 127,
      "red": 127,
    },
  ],
]
`;

exports[`should maintain state between frames 1`] = `
Array [
  Array [
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
  ],
  Array [
    Object {
      "alpha": 1,
      "blue": 255,
      "green": 255,
      "red": 255,
    },
    Object {
      "alpha": 1,
      "blue": 255,
      "green": 255,
      "red": 255,
    },
    Object {
      "alpha": 1,
      "blue": 255,
      "green": 255,
      "red": 255,
    },
    Object {
      "alpha": 1,
      "blue": 255,
      "green": 255,
      "red": 255,
    },
  ],
]
`;

exports[`should normalize color values: using a color name 1`] = `
Array [
  Object {
    "alpha": 1,
    "blue": 54,
    "green": 65,
    "red": 255,
  },
]
`;

exports[`should normalize color values: using a complete RGBA object 1`] = `
Array [
  Object {
    "alpha": 0,
    "blue": 76,
    "green": 51,
    "red": 25,
  },
]
`;

exports[`should normalize color values: using a hex code 1`] = `
Array [
  Object {
    "alpha": 1,
    "blue": 0,
    "green": 0,
    "red": 255,
  },
]
`;

exports[`should normalize color values: using a number 1`] = `
Array [
  Object {
    "alpha": 1,
    "blue": 127,
    "green": 127,
    "red": 127,
  },
]
`;

exports[`should normalize color values: using a partial RGBA object 1`] = `
Array [
  Object {
    "alpha": 1,
    "blue": 0,
    "green": 0,
    "red": 255,
  },
]
`;

exports[`should normalize color values: using null 1`] = `
Array [
  Object {
    "alpha": 0,
    "blue": 0,
    "green": 0,
    "red": 0,
  },
]
`;

exports[`should wrap when referencing out of bounds cell 1`] = `
Array [
  Array [
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 127,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 127,
    },
  ],
  Array [
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 127,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 127,
      "red": 0,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 127,
    },
    Object {
      "alpha": 1,
      "blue": 0,
      "green": 0,
      "red": 0,
    },
  ],
]
`;
